libs/server/Storage/Session/MainStore/MainStoreOps.cs

```diff
@@ -618,32 +618,171 @@ public unsafe GarnetStatus RENAME(ArgSlice oldKeySlice, ArgSlice newKeySlice, St
                         var valObj = value.garnetObject;
                         byte[] newKeyArray = newKeySlice.ToArray();
 
+                        // valObj already has expiration time, so no need to write expiration logic here
+                        SET(newKeyArray, valObj, ref objectContext);
+
+                        // Delete the old key
+                        DELETE(oldKeyArray, StoreType.Object, ref context, ref objectContext);
+
+                        returnStatus = GarnetStatus.OK;
+                    }
+                }
+                finally
+                {
+                    if (createTransaction)
+                        txnManager.Commit(true);
+                }
+            }
+
+            return returnStatus;
+        }
+
+        /// <summary>
+        /// Renames key to newkey if newkey does not yet exist. It returns an error when key does not exist.
+        /// </summary>
+        /// <param name="oldKeySlice">The old key to be renamed.</param>
+        /// <param name="newKeySlice">The new key name.</param>
+        /// <param name="storeType">The type of store to perform the operation on.</param>
+        /// <returns>
+        ///     GarnetStatus OK means rename operation is successful
+        ///     GarnetStatus NOTFOUND means old key does not exists
+        ///     GarnetStatus MOVED means new key already exists
+        /// </returns>
+        public unsafe GarnetStatus RENAMENX(ArgSlice oldKeySlice, ArgSlice newKeySlice, StoreType storeType)
```

Comment:
Since this is majorly similar to RENAME, I would suggest adding a bool parameter to the RENAME method to specify if rename should happen only if key doesn't exist, instead of copying all this code (as any future change to this would need to happen in multiple locations).

Reply:
Agree I will make it into a single method, was planned but didn't do it at the end

Reply:
Converted into single method. Not my best work but I think its good enough.